diff --git a/autorandr.py b/autorandr.py
index 25b0f53..2cc887c 100755
--- a/autorandr.py
+++ b/autorandr.py
@@ -770,9 +770,10 @@ def apply_configuration(new_configuration, current_configuration, dry_run=False)
 
     fb_dimensions = get_fb_dimensions(new_configuration)
     try:
-        base_argv += ["--fb", "%dx%d" % fb_dimensions]
+        fb_argv += ["--fb", "%dx%d" % fb_dimensions]
     except:
         # Failed to obtain frame-buffer size. Doesn't matter, xrandr will choose for the user.
+        fb_argv = []
         pass
 
     auxiliary_changes_pre = []
@@ -852,7 +853,11 @@ def apply_configuration(new_configuration, current_configuration, dry_run=False)
     # Enable the remaining outputs in pairs of two operations
     operations = disable_outputs + enable_outputs
     for index in range(0, len(operations), 2):
-        argv = base_argv + list(chain.from_iterable(operations[index:index + 2]))
+        argv = base_argv.copy()
+        if index >= len(disable_outputs):
+            argv += fb_argv
+        argv += list(chain.from_iterable(operations[index:index + 2]))
+        
         if call_and_retry(argv, dry_run=dry_run) != 0:
             raise AutorandrException("Command failed: %s" % " ".join(argv))
 
